core/src/main/java/io/parsingdata/metal/expression/value/FoldRight.java

```diff
@@ -16,37 +16,38 @@
 
 package io.parsingdata.metal.expression.value;
 
-import static io.parsingdata.metal.Shorthand.con;
-import static io.parsingdata.metal.Util.checkNotNull;
-
 import io.parsingdata.metal.data.Environment;
-import io.parsingdata.metal.data.ParseValueList;
+import io.parsingdata.metal.data.OptionalValueList;
 import io.parsingdata.metal.encoding.Encoding;
 
+import static io.parsingdata.metal.Shorthand.con;
+import static io.parsingdata.metal.Util.checkNotNull;
+
 public class FoldRight implements ValueExpression {
 
-    private final String _name;
+    private final ValueExpression _values;
     private final Reducer _reducer;
     private final ValueExpression _init;
 
-    public FoldRight(final String name, final Reducer reducer, final ValueExpression init) {
-        _name = checkNotNull(name, "name");
+    public FoldRight(final ValueExpression values, final Reducer reducer, final ValueExpression init) {
+        _values = checkNotNull(values, "values");
         _reducer = checkNotNull(reducer, "reducer");
         _init = init;
     }
 
     @Override
-    public OptionalValue eval(final Environment env, final Encoding enc) {
-        final OptionalValue init = _init != null ? _init.eval(env, enc) : OptionalValue.empty();
-        final ParseValueList values = env.order.getAll(_name);
-        if (values.isEmpty()) { return init; }
-        if (init.isPresent()) { return fold(env, enc, _reducer, init, values); }
-        return fold(env, enc, _reducer, OptionalValue.of(values.head), values.tail);
+    public OptionalValueList eval(final Environment env, final Encoding enc) {
+        final OptionalValueList init = _init != null ? _init.eval(env, enc) : OptionalValueList.EMPTY;
+        if (init.size > 1) { throw new RuntimeException("Init may not contain more than a single value."); }
+        final OptionalValueList values = _values.eval(env, enc);
+        if (values.isEmpty() || values.containsEmpty()) { return init; }
+        if (!init.isEmpty()) { return OptionalValueList.create(fold(env, enc, _reducer, init.head, values)); }
+        return OptionalValueList.create(fold(env, enc, _reducer, values.head, values.tail));
     }
 
-    private OptionalValue fold(final Environment env, final Encoding enc, final Reducer reducer, final OptionalValue head, final ParseValueList tail) {
+    private OptionalValue fold(final Environment env, final Encoding enc, final Reducer reducer, final OptionalValue head, final OptionalValueList tail) {
         if (!head.isPresent() || tail.isEmpty()) { return head; }
-        return fold(env, enc, reducer, reducer.reduce(con(tail.head), con(head.get())).eval(env, enc), tail.tail);
+        return fold(env, enc, reducer, reducer.reduce(con(tail.head.get()), con(head.get())).eval(env, enc).head, tail.tail);
```

Comment:
(See my earlier comment about only passing on the `head` into the next `reduce`.)